internal/controller/suite_test.go

```diff
@@ -27,33 +31,52 @@ import (
 
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/envtest"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 
+	numaflowv1 "github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
 	apiv1 "github.com/numaproj/numaplane/pkg/apis/numaplane/v1alpha1"
 )
 
 // These tests use Ginkgo (BDD-style Go testing framework). Refer to
 // http://onsi.github.io/ginkgo/ to learn more about Ginkgo.
 
-var cfg *rest.Config
-var k8sClient client.Client
-var testEnv *envtest.Environment
+var (
+	cfg             *rest.Config
+	k8sClient       client.Client
+	testEnv         *envtest.Environment
+	externalCRDsDir string
+)
 
 func TestControllers(t *testing.T) {
 	RegisterFailHandler(Fail)
 
-	RunSpecs(t, "Controller Suite")
+	RunSpecs(t, "Controllers Suite")
 }
 
 var _ = BeforeSuite(func() {
 	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
 
+	// Download Numaflow CRDs
+	crdsURLs := []string{
+		"https://raw.githubusercontent.com/numaproj/numaflow/main/config/base/crds/minimal/numaflow.numaproj.io_interstepbufferservices.yaml",
+		"https://raw.githubusercontent.com/numaproj/numaflow/main/config/base/crds/minimal/numaflow.numaproj.io_pipelines.yaml",
+		"https://raw.githubusercontent.com/numaproj/numaflow/main/config/base/crds/minimal/numaflow.numaproj.io_vertices.yaml",
+	}
+	externalCRDsDir = filepath.Join("..", "..", "config", "crd", "external")
+	for _, crdURL := range crdsURLs {
+		downloadCRD(crdURL, externalCRDsDir)
+	}
+
 	By("bootstrapping test environment")
+	// TODO: IDEA: could set useExistingCluster via an env var to be able to reuse some tests cases in e2e tests
+	// by setting this variable in CI for unit tests and e2e tests appropriately.
```

Comment:
what does it mean "to reuse some test cases in e2e tests"? do you have an example of what you mean?

Reply:
The test cases in the `pipelinerollout_controller_test.go` file test PipelineRollout lifecycle (create, update, delete operations). These could be reused in e2e tests and, when using a real cluster, also the deletion of the Numaflow Pipeline on deletion of PipelineRollout could be tested (https://github.com/numaproj/numaplane/pull/41/files#diff-31f7e2289e4cc3f50a723cc05922642d3cc20247ed71f66dc7dab1a73d3f302cR237-R245).

Reply:
Interesting. Possibly. Although, that will likely use the same test framework "github.com/stretchr/testify/suite" we were previously using....so you'd have to see if it would be easily callable from there, or otherwise may not be worth it